File: 01436. Destination City/solution.cpp
/**
 * Problem: 1436. Destination City
 *
 * Difficulty: Easy
 *
 * Language: C++
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

class Solution
{
public:
  string destCity(vector<vector<string>> &paths)
  {
    // Set to store all destination cities
    unordered_set<string> destinations;

    // Add all destination cities to the set
    for (const auto &path : paths)
      destinations.insert(path[1]);

    // Remove cities that also appear as sources (have outgoing paths)
    for (const auto &path : paths)
    {
      // Get source city from current path
      string sourceCity = path[0];

      // If source exists in destinations, it's not a final destination
      if (destinations.find(sourceCity) != destinations.end())
        destinations.erase(sourceCity);
    }

    // The remaining city in the set is the final destination
    return *destinations.begin();
  }
};

File: 01436. Destination City/solution.ts
/**
 * Problem: 1436. Destination City
 *
 * Difficulty: Easy
 *
 * Language: TypeScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Finds the destination city that is not a starting point
 *
 * @param paths - Array of [start, destination] pairs
 *
 * @returns The destination city with no outgoing path
 */
const destCity = (paths: string[][]): string => {
  // Set to store all starting cities
  const startingCities: Set<string> = new Set()

  // Add all starting cities to the set
  for (const path of paths) startingCities.add(path[0])

  // Check each destination to find one that is not a starting city
  for (const path of paths) {
    // Get destination city from current path
    const destination: string = path[1]

    // If destination is not in starting cities, it's the answer
    if (!startingCities.has(destination)) return destination
  }

  // Return empty string if no destination found
  return ''
}
